rust-runtime/aws-smithy-http/src/byte_stream/bytestream_util.rs

```diff
@@ -236,11 +301,290 @@ impl Body for PathBody {
+        // read a little bit from one of the clones
+        let some_data = body1
+            .data()
+            .await
+            .expect("should have some data")
+            .expect("read should not fail");
+        // The size of one read should be equal to that of the buffer size
+        assert_eq!(some_data.len(), 16384);
+
+        assert_eq!(
+            ByteStream::new(body1).collect().await?.remaining(),
+            file_length as usize - some_data.len()
+        );
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn fsbuilder_length_is_used_as_size_hint() -> Result<(), Box<dyn std::error::Error>> {
+        let mut file = NamedTempFile::new()?;
+        write!(
+            file,
+            "A very long sentence that's clearly longer than a single byte."
+        )?;
+        // Ensure that the file was written to
+        file.flush().expect("flushing is OK");
+
+        let body = FsBuilder::new()
+            .path(&file)
+            // The file is longer than 1 byte, let's see if this is used to generate the size hint
+            .length(Length::Exact(1))
+            .build()
+            .await?
+            .into_inner();
+
+        assert_eq!(body.size_hint().exact(), Some(1));
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn fsbuilder_respects_length() -> Result<(), Box<dyn std::error::Error>> {
+        let mut file = NamedTempFile::new()?;
+        let line_0 = "Line 0\n";
+        let line_1 = "Line 1\n";
+
+        write!(file, "{}", line_0)?;
+        write!(file, "{}", line_1)?;
+
+        // Ensure that the file was written to
+        file.flush().expect("flushing is OK");
+
+        let body = FsBuilder::new()
+            .path(&file)
+            // We're going to read line 0 only
+            .length(Length::Exact(line_0.len() as u64))
+            .build()
+            .await?;
+
+        let data = body.collect().await?.into_bytes();
+        let data_str = String::from_utf8(data.to_vec())?;
+
+        assert_eq!(&data_str, line_0);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn fsbuilder_supports_offset() -> Result<(), Box<dyn std::error::Error>> {
+        let mut file = NamedTempFile::new()?;
+        let line_0 = "Line 0\n";
+        let line_1 = "Line 1\n";
+
+        write!(file, "{}", line_0)?;
+        write!(file, "{}", line_1)?;
+
+        // Ensure that the file was written to
+        file.flush().expect("flushing is OK");
+
+        let body = FsBuilder::new()
+            .path(&file)
+            // We're going to skip the first line by using offset
+            .offset(line_0.len() as u64)
+            .build()
+            .await?;
+
+        let data = body.collect().await?.into_bytes();
+        let data_str = String::from_utf8(data.to_vec())?;
+
+        assert_eq!(&data_str, line_1);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn fsbuilder_offset_and_length_work_together() -> Result<(), Box<dyn std::error::Error>> {
```

Comment:
Nit: I think tests that `unwrap` are a lot easier to debug when things go wrong than ones that return an error. The latter doesn't give you a nice line number to work with (unless I'm missing some neat trick).

Reply:
yeah you'd need `context!()`